junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/StoredValues.java

```diff
@@ -0,0 +1,153 @@
+
+package org.junit.jupiter.api.extension;
+
+import static org.junit.platform.commons.meta.API.Usage.Experimental;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+import org.junit.jupiter.api.extension.ExtensionContext.Store;
+import org.junit.platform.commons.meta.API;
+
+/**
+ * Provides isolated, type-safe access to values in a {@link Store}.
+ *
+ * <p>Example use:
+ *
+ * <pre>{@code
+ * public class TimingExtension
+ *     implements BeforeTestExecutionCallback, AfterTestExecutionCallback {
+ *   private static final StoredValues<Method, Instant> START_TIMES
+ *       = new StoredValues<>();
+ * 
+ *   @Override
+ *   public void beforeTestExecution(ExtensionContext context) {
+ *     START_TIMES.put(
+ *         getStore(context), context.getRequiredTestMethod(), Instant.now());
+ *   }
+ * }
+ * }</pre>
```

Comment:
With the current API, each time to you access a Store for a key, you have to specify the expected value, which is error-prone.

I see no problem with having multiple `StoredValues` values. It's like having multiple fields. Or a `ThreadLocal`.

Plus you are not required to use the new API.